phoenix-core/src/main/java/org/apache/phoenix/compile/CreateIndexCompiler.java

```diff
@@ -45,11 +59,92 @@ public CreateIndexCompiler(PhoenixStatement statement, Operation operation) {
+        if (indexWhere == null) {
+            return;
+        }
+        PhoenixConnection connection = context.getConnection();
+        IndexWhereParseNodeVisitor indexWhereParseNodeVisitor = new IndexWhereParseNodeVisitor();
+        indexWhere.accept(indexWhereParseNodeVisitor);
+        if (indexWhereParseNodeVisitor.hasSubquery) {
+            throw new SQLExceptionInfo.Builder(SQLExceptionCode.INVALID_INDEX_WHERE_WITH_SUBQUERY).
+                    build().buildException();
+        }
+        ExpressionCompiler expressionCompiler = new ExpressionCompiler(context);
+        Expression indexWhereExpression = indexWhere.accept(expressionCompiler);
+        PTable dataTable = PhoenixRuntime.getTable(connection, dataTableName.toString());
+
+        boolean autoCommit = connection.getAutoCommit();
+        connection.setAutoCommit(false);
+        StringBuilder stringBuilder = new StringBuilder("UPSERT INTO ");
+        stringBuilder.append(dataTableName);
+        stringBuilder.append(" Values(");
+        int i = dataTable.getBucketNum() != null ? 1 : 0;
+        for (; i < dataTable.getColumns().size() - 1; i++) {
+            PDataType dataType = dataTable.getColumns().get(i).getDataType();
+            stringBuilder.append(getValue(dataType) + ",");
+        }
+        PDataType dataType = dataTable.getColumns().get(i).getDataType();
+        stringBuilder.append(getValue(dataType)+ ")");
+        PreparedStatement ps = context.getConnection().prepareStatement(stringBuilder.toString());
+        ps.execute();
+        Iterator<Pair<byte[],List<Cell>>> dataTableNameAndMutationIterator =
+                PhoenixRuntime.getUncommittedDataIterator(connection);
+        Pair<byte[], List<Cell>> dataTableNameAndMutation = null;
+        while (dataTableNameAndMutationIterator.hasNext()) {
+            dataTableNameAndMutation = dataTableNameAndMutationIterator.next();
+            if (!java.util.Arrays.equals(dataTableNameAndMutation.getFirst(),
+                    dataTableName.toString().getBytes())) {
+                dataTableNameAndMutation = null;
+            }
+        }
+        if (dataTableNameAndMutation == null) {
+            ps.close();
+            connection.setAutoCommit(autoCommit);
+            throw new RuntimeException("Unexpected result from " +
+                    "PhoenixRuntime#getUncommittedDataIterator for " + dataTableName);
+        }
+        ImmutableBytesWritable ptr = new ImmutableBytesWritable();
+        ptr.set(ByteUtil.EMPTY_BYTE_ARRAY);
+        MultiKeyValueTuple tuple = new MultiKeyValueTuple(dataTableNameAndMutation.getSecond());
+        if (!indexWhereExpression.evaluate(tuple, ptr)) {
+            ps.close();
+            connection.setAutoCommit(autoCommit);
+            throw new SQLExceptionInfo.Builder(SQLExceptionCode.CANNOT_EVALUATE_INDEX_WHERE).
+                    build().buildException();
+        }
+        ps.close();
+        connection.setAutoCommit(autoCommit);
```

Comment:
would it be good put this code in a try/finally block?

Reply:
Yes

Reply:
Done